kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/IcebergWriterFactory.java

```diff
@@ -0,0 +1,112 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.connect.data;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.connect.IcebergSinkConfig;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.types.Types.StructType;
+import org.apache.iceberg.util.Tasks;
+import org.apache.kafka.connect.errors.DataException;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class IcebergWriterFactory {
+
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergWriterFactory.class);
+
+  private final Catalog catalog;
+  private final IcebergSinkConfig config;
+
+  public IcebergWriterFactory(Catalog catalog, IcebergSinkConfig config) {
+    this.catalog = catalog;
+    this.config = config;
+  }
+
+  public RecordWriter createWriter(
+      String tableName, SinkRecord sample, boolean ignoreMissingTable) {
+    TableIdentifier identifier = TableIdentifier.parse(tableName);
+    Table table;
+    try {
+      table = catalog.loadTable(identifier);
+    } catch (NoSuchTableException nst) {
+      if (config.autoCreateEnabled()) {
+        table = autoCreateTable(tableName, sample);
+      } else if (ignoreMissingTable) {
+        return new NoOpWriter();
+      } else {
+        throw nst;
+      }
+    }
+
+    return new IcebergWriter(table, tableName, config);
+  }
+
+  @VisibleForTesting
+  Table autoCreateTable(String tableName, SinkRecord sample) {
+    StructType structType;
+    if (sample.valueSchema() == null) {
+      structType =
+          SchemaUtils.inferIcebergType(sample.value(), config)
+              .orElseThrow(() -> new DataException("Unable to create table from empty object"))
+              .asStructType();
+    } else {
+      structType = SchemaUtils.toIcebergType(sample.valueSchema(), config).asStructType();
+    }
+
+    org.apache.iceberg.Schema schema = new org.apache.iceberg.Schema(structType.fields());
+    TableIdentifier identifier = TableIdentifier.parse(tableName);
+
+    List<String> partitionBy = config.tableConfig(tableName).partitionBy();
+    PartitionSpec spec;
+    try {
+      spec = SchemaUtils.createPartitionSpec(schema, partitionBy);
+    } catch (Exception e) {
+      LOG.error(
+          "Unable to create partition spec {}, table {} will be unpartitioned",
+          partitionBy,
+          identifier,
+          e);
+      spec = PartitionSpec.unpartitioned();
+    }
```

Comment:
My thought was to have the sink be more permissive in this case, i.e. if the sink is fanning out to several different tables, don't error out if one of them can't be partitioned as that can be difficult to recover from. There is room for improvement in error handling to help with recovery, e.g. adding in DLQ support for records that can't be processed.

Reply:
I agree that we don't want to fail if the partitioning doesn't work. For example, if you expect events to have a `event_ts` column but it isn't in all event types. We should still try to make progress rather than ignoring the events or (worse) failing operation.